### PR TITLE
Fix space logic after the emoji

### PR DESCRIPTION
**Pull request by Puneet-here**
### Details

We only wanna add space after emoji when user uses an emoji keyword (example: `:cry:`) to add an emoji at mobile.
### Fixed Issues

$ https://github.com/Expensify/App/issues/13343
PROPOSAL: https://github.com/Expensify/App/issues/13343#issuecomment-1339011328

### Tests

**Desktop, Web**
1. Go to a chat
2. Open emoji picker
3. Choose an emoji
4. Verify no space gets added after the emoji
5. At composer write `:cry:`
6. Verify no space gets added after the emoji
7. Send a message and edit it
8. Repeat all the steps above for edit composer

---
**Android, ios, mWeb**
1. Go to a chat
2. Open emoji picker
3. Choose an emoji
5. Verify no space gets added after the emoji
6. At composer write `:cry:`
7. Verify space gets added after the emoji
8. Send a message and edit it
9. Repeat all the steps above for edit composer
- [x] Verify that no errors appear in the JS console

### Offline tests

None
### QA Steps

**Desktop, Web**
1. Go to a chat
2. Open emoji picker
3. Choose an emoji
4. Verify no space gets added after the emoji
5. At composer write `:cry:`
6. Verify no space gets added after the emoji
7. Send a message and edit it
8. Repeat all the steps above for edit composer

---
**Android, ios, mWeb**
1. Go to a chat
2. Open emoji picker
3. Choose an emoji
5. Verify no space gets added after the emoji
6. At composer write `:cry:`
7. Verify space gets added after the emoji
8. Send a message and edit it
8. Repeat all the steps above for edit composer
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/101883770/209995802-1e82668f-be66-45c4-b770-c8bc54e18d1b.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/101883770/209995709-5234eadf-608c-4bad-8792-308397b0a712.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/101883770/209996113-7a8c7abc-4d6b-46b9-b668-0c224b9407b1.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/101883770/209995793-4f268e75-f5db-4da3-8cb9-de58d9119174.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/101883770/209995781-27ecfc46-2c76-4354-b514-932bd0eee588.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/101883770/209995723-4fd4b172-c3e3-4473-9932-6faa2d3e8754.mov

</details>

### CODE REVIEW

**Comment by thesahindia** [comment]
@Puneet-here, 1 test failed. Please look into it.

**Comment by Puneet-here** [comment]
@thesahindia, updated!

**Comment by thesahindia** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/83179295/210012929-aff11220-01f9-4b0f-bffb-59d6d12bb6f0.mov





</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/83179295/210012934-0896bce1-e49d-49d8-b9e4-b73ee0904473.mov




</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/83179295/210012943-5b61f2bf-79dd-46fe-a1cd-817629d73977.mov





</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/83179295/210012947-c5f2fd5f-e65f-40a8-a935-2a7f23c10150.mov





</details>

<details>

<summary>iOS</summary>

https://user-images.githubusercontent.com/83179295/210012958-9789f66b-8b48-419f-bb05-6542c1c6885d.mov




</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/83179295/210012979-eabb4627-3f58-416b-9c57-beb48413828f.mov




</details>

**Review by thesahindia — DISMISSED** [review]
Looks good and works well!

C+ reviewed 🎀👀🎀

All yours @youssef-lr

File: src/libs/EmojiUtils.js [hunk 1]
```diff
@@ -186,9 +186,10 @@ function addToFrequentlyUsedEmojis(frequentlyUsedEmojis, newEmoji) {
 /**
  * Replace any emoji name in a text with the emoji icon
  * @param {String} text
+ * @param {Boolean} addSpace
  * @returns {String}
  */
-function replaceEmojis(text) {
+function replaceEmojis(text, addSpace = false) {
```
Comment (youssef-lr) [inline src/libs/EmojiUtils.js:192]: I think we should rename this parameter to `isSmallScreenWidth` for more clarity. As in the tests below passing `addSpace` as true when there's text after it is misleading, another engineer might assume that we always add space if that parameter is true.

**Review by youssef-lr — CHANGES_REQUESTED** [review]
Looks good! Left a minor comment.

File: tests/unit/EmojiTest.js [hunk 2]
```diff
@@ -78,14 +78,19 @@ describe('EmojiTest', () => {
         expect(EmojiUtils.containsOnlyEmojis('😄  👋')).toBe(true);
     });
 
-    it('replaces an emoji code with an emoji and a space', () => {
+    it('replaces an emoji code with an emoji and a space if addSpace is passed as true', () => {
```
Comment (youssef-lr) [inline tests/unit/EmojiTest.js:81]:
```suggestion
    it('replaces an emoji code with an emoji and a space on mobile', () => {
```

File: tests/unit/EmojiTest.js [hunk 3]
```diff
@@ -78,14 +78,19 @@ describe('EmojiTest', () => {
...
         const text = 'Hi :smile:';
-        expect(EmojiUtils.replaceEmojis(text)).toBe('Hi 😄 ');
+        expect(EmojiUtils.replaceEmojis(text, true)).toBe('Hi 😄 ');
     });
 
-    it('will not add a space after the last emoji if there is text after it', () => {
+    it('will not add a space after the last emoji when there is text after it even if addSpace is passed as true', () => {
```
Comment (youssef-lr) [inline tests/unit/EmojiTest.js:86]:
```suggestion
    it('will not add a space after the last emoji when there is text after it', () => {
```
We can keep this one as is. The purpose here is to test that we're not adding a space if there's text after the emoji, regardless of which device we're on.

File: tests/unit/EmojiTest.js [hunk 4]
```diff
@@ -78,14 +78,19 @@ describe('EmojiTest', () => {
...
         const text = 'Hi :smile::wave:no space after last emoji';
-        expect(EmojiUtils.replaceEmojis(text)).toBe('Hi 😄👋no space after last emoji');
+        expect(EmojiUtils.replaceEmojis(text, true)).toBe('Hi 😄👋no space after last emoji');
```
Comment (youssef-lr) [inline tests/unit/EmojiTest.js:88]:
```suggestion
        expect(EmojiUtils.replaceEmojis(text)).toBe('Hi 😄👋no space after last emoji');
```
Let's keep this test as is. And add another one:

`expect(will not add a space after the last emoji when there is text after it on mobile)` which adds `true` in the call.

File: tests/unit/EmojiTest.js [hunk 5]
```diff
@@ -78,14 +78,19 @@ describe('EmojiTest', () => {
...
+    });
+
+    it('will not add a space after the last emoji if addSpace is not true', () => {
```
Comment (youssef-lr) [inline tests/unit/EmojiTest.js:91]:
```suggestion
    it('will not add a space after the last emoji if we\'re not on mobile', () => {
```

File: src/libs/EmojiUtils.js [hunk 6]
```diff
@@ -186,9 +186,10 @@ function addToFrequentlyUsedEmojis(frequentlyUsedEmojis, newEmoji) {
 /**
  * Replace any emoji name in a text with the emoji icon
```
Comment (youssef-lr) [inline src/libs/EmojiUtils.js:187]:
```suggestion
 * Replace any emoji name in a text with the emoji icon.
 * If we're on mobile, we also add a space after the emoji granted there's no text after it.
```

**Comment by Puneet-here** [comment]
Done, updated! @youssef-lr @thesahindia

**Review by youssef-lr — APPROVED** [review]
LGTM! Thanks for the changes.

**Pull request merged**